File: vertx-auth-mongo/src/main/java/io/vertx/ext/auth/mongo/MongoAuth.java
/*
 * Copyright 2014 Red Hat, Inc.
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * and Apache License v2.0 which accompanies this distribution.
 *
 * The Eclipse Public License is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * The Apache License v2.0 is available at
 * http://www.opensource.org/licenses/apache2.0.php
 *
 * You may elect to redistribute this code under either of these licenses.
 */

package io.vertx.ext.auth.mongo;

import io.vertx.codegen.annotations.Fluent;
import io.vertx.codegen.annotations.VertxGen;
import io.vertx.core.AsyncResult;
import io.vertx.core.Handler;
import io.vertx.core.json.JsonObject;
import io.vertx.ext.auth.AuthProvider;
import io.vertx.ext.auth.User;
import io.vertx.ext.auth.mongo.impl.MongoAuthImpl;
import io.vertx.ext.mongo.MongoClient;

import java.util.List;

/**
 * An extension of AuthProvider which is using {@link MongoClient} as store
 *
 * @author mremme
 */
@VertxGen
public interface MongoAuth extends AuthProvider {

  /**
   * The property name to be used to set the name of the collection inside the config
   */
  String PROPERTY_COLLECTION_NAME = "collectionName";

  /**
   * The property name to be used to set the name of the field, where the username is stored inside
   */
  String PROPERTY_USERNAME_FIELD = "usernameField";

  /**
   * The property name to be used to set the name of the field, where the roles are stored inside
   */
  String PROPERTY_ROLE_FIELD = "roleField";

  /**
   * The property name to be used to set the name of the field, where the permissions are stored inside
   */
  String PROPERTY_PERMISSION_FIELD = "permissionField";

  /**
   * The property name to be used to set the name of the field, where the password is stored inside
   */
  String PROPERTY_PASSWORD_FIELD = "passwordField";

  /**
   * The property name to be used to set the name of the field, where the username for the credentials is stored inside
   */
  String PROPERTY_CREDENTIAL_USERNAME_FIELD = "usernameCredentialField";

  /**
   * The property name to be used to set the name of the field, where the password for the credentials is stored inside
   */
  String PROPERTY_CREDENTIAL_PASSWORD_FIELD = "passwordCredentialField";

  /**
   * The property name to be used to set the name of the field, where the SALT is stored inside
   */
  String PROPERTY_SALT_FIELD = "saltField";

  /**
   * The property name to be used to set the name of the field, where the salt style is stored inside
   *
   * @see HashSaltStyle
   */
  String PROPERTY_SALT_STYLE = "saltStyle";

  /**
   * The default name of the collection to be used
   */
  String DEFAULT_COLLECTION_NAME = "user";

  /**
   * The default name of the property for the username, like it is stored in mongodb
   */
  String DEFAULT_USERNAME_FIELD = "username";

  /**
   * The default name of the property for the password, like it is stored in mongodb
   */
  String DEFAULT_PASSWORD_FIELD = "password";

  /**
   * The default name of the property for the roles, like it is stored in mongodb. Roles are expected to be saved as
   * JsonArray
   */
  String DEFAULT_ROLE_FIELD = "roles";

  /**
   * The default name of the property for the permissions, like it is stored in mongodb. Permissions are expected to be
   * saved as JsonArray
   */
  String DEFAULT_PERMISSION_FIELD = "permissions";

  /**
   * The default name of the property for the username, like it is transported in credentials by method
   * {@link #authenticate(JsonObject, Handler)}
   */
  String DEFAULT_CREDENTIAL_USERNAME_FIELD = DEFAULT_USERNAME_FIELD;

  /**
   * The default name of the property for the password, like it is transported in credentials by method
   * {@link #authenticate(JsonObject, Handler)}
   */
  String DEFAULT_CREDENTIAL_PASSWORD_FIELD = DEFAULT_PASSWORD_FIELD;

  /**
   * The default name of the property for the salt field
   */
  String DEFAULT_SALT_FIELD = "salt";

  /**
   * The prefix which is used by the method {@link User#isAuthorised(String, Handler)} when checking for role access
   */
  String ROLE_PREFIX = "role:";

  /**
   * Creates an instance of MongoAuth by using the given {@link MongoClient} and configuration object. An example for a
   * configuration object:
   *
   * <pre>
   * JsonObject js = new JsonObject();
   * js.put(MongoAuth.PROPERTY_COLLECTION_NAME, createCollectionName(MongoAuth.DEFAULT_COLLECTION_NAME));
   * </pre>
   *
   * @param mongoClient
   *          an instance of {@link MongoClient} to be used for data storage and retrival
   * @param config
   *          the configuration object for the current instance. By this
   * @return the created instance of {@link MongoAuth}s
   */
  static MongoAuth create(MongoClient mongoClient, JsonObject config) {
    return new MongoAuthImpl(mongoClient, config);
  }

  /**
   * Set the name of the collection to be used. Defaults to {@link #DEFAULT_COLLECTION_NAME}
   *
   * @param collectionName
   *          the name of the collection to be used for storing and reading user data
   * @return the current instance itself for fluent calls
   */
  @Fluent
  MongoAuth setCollectionName(String collectionName);

  /**
   * Set the name of the field to be used for the username. Defaults to {@link #DEFAULT_USERNAME_FIELD}
   *
   * @param fieldName
   *          the name of the field to be used
   * @return the current instance itself for fluent calls
   */
  @Fluent
  MongoAuth setUsernameField(String fieldName);

  /**
   * Set the name of the field to be used for the password Defaults to {@link #DEFAULT_PASSWORD_FIELD}
   *
   * @param fieldName
   *          the name of the field to be used
   * @return the current instance itself for fluent calls
   */
  @Fluent
  MongoAuth setPasswordField(String fieldName);

  /**
   * Set the name of the field to be used for the roles. Defaults to {@link #DEFAULT_ROLE_FIELD}. Roles are expected to
   * be saved as JsonArray
   *
   * @param fieldName
   *          the name of the field to be used
   * @return the current instance itself for fluent calls
   */
  @Fluent
  MongoAuth setRoleField(String fieldName);

  /**
   * Set the name of the field to be used for the permissions. Defaults to {@link #DEFAULT_PERMISSION_FIELD}.
   * Permissions are expected to be saved as JsonArray
   *
   * @param fieldName
   *          the name of the field to be used
   * @return the current instance itself for fluent calls
   */
  @Fluent
  MongoAuth setPermissionField(String fieldName);

  /**
   * Set the name of the field to be used as property for the username in the method
   * {@link #authenticate(JsonObject, io.vertx.core.Handler)}. Defaults to {@link #DEFAULT_CREDENTIAL_USERNAME_FIELD}
   *
   * @param fieldName
   *          the name of the field to be used
   * @return the current instance itself for fluent calls
   */
  @Fluent
  MongoAuth setUsernameCredentialField(String fieldName);

  /**
   * Set the name of the field to be used as property for the password of credentials in the method
   * {@link #authenticate(JsonObject, io.vertx.core.Handler)}. Defaults to {@link #DEFAULT_CREDENTIAL_PASSWORD_FIELD}
   *
   * @param fieldName
   *          the name of the field to be used
   * @return the current instance itself for fluent calls
   */
  @Fluent
  MongoAuth setPasswordCredentialField(String fieldName);

  /**
   * Set the name of the field to be used for the salt. Only used when {@link HashStrategy#setSaltStyle(HashSaltStyle)} is
   * set to {@link HashSaltStyle#COLUMN}
   *
   * @param fieldName
   *          the name of the field to be used
   * @return the current instance itself for fluent calls
   */
  @Fluent
  MongoAuth setSaltField(String fieldName);

  /**
   * The name of the collection used to store User objects inside. Defaults to {@link #DEFAULT_COLLECTION_NAME}
   *
   * @return the collectionName
   */
  String getCollectionName();

  /**
   * Get the name of the field to be used for the username. Defaults to {@link #DEFAULT_USERNAME_FIELD}
   *
   * @return the usernameField
   */
  String getUsernameField();

  /**
   * Get the name of the field to be used for the password Defaults to {@link #DEFAULT_PASSWORD_FIELD}
   *
   * @return the passwordField
   */
  String getPasswordField();

  /**
   * Get the name of the field to be used for the roles. Defaults to {@link #DEFAULT_ROLE_FIELD}. Roles are expected to
   * be saved as JsonArray
   *
   * @return the roleField
   */
  String getRoleField();

  /**
   * Get the name of the field to be used for the permissions. Defaults to {@link #DEFAULT_PERMISSION_FIELD}.
   * Permissions are expected to be saved as JsonArray
   *
   * @return the permissionField
   */
  String getPermissionField();

  /**
   * Get the name of the field to be used as property for the username in the method
   * {@link #authenticate(JsonObject, io.vertx.core.Handler)}. Defaults to {@link #DEFAULT_CREDENTIAL_USERNAME_FIELD}
   *
   * @return the usernameCredentialField
   */
  String getUsernameCredentialField();

  /**
   * Get the name of the field to be used as property for the password of credentials in the method
   * {@link #authenticate(JsonObject, io.vertx.core.Handler)}. Defaults to {@link #DEFAULT_CREDENTIAL_PASSWORD_FIELD}
   *
   * @return the passwordCredentialField
   */
  String getPasswordCredentialField();

  /**
   * Get the name of the field to be used for the salt. Only used when {@link HashStrategy#setSaltStyle(HashSaltStyle)} is
   * set to {@link HashSaltStyle#COLUMN}
   *
   * @return the saltField
   */
  String getSaltField();

  /**
   * The HashStrategy which is used by the current instance
   *
   * @param hashStrategy
   *          the {@link HashStrategy} to be set
   * @return the current instance itself for fluent calls
   *
   */
  @Fluent
  MongoAuth setHashStrategy(HashStrategy hashStrategy);

  /**
   * The HashStrategy which is used by the current instance
   *
   * @return the defined instance of {@link HashStrategy}
   */
  HashStrategy getHashStrategy();

  /**
   * The Hash Algorithm which is used by the current instance
   *
   * @param hashAlgorithm
   *          the {@link HashAlgorithm} to be set
   * @return the current instance itself for fluent calls
   *
   */
  @Fluent
  MongoAuth setHashAlgorithm(HashAlgorithm hashAlgorithm);

  /**
   * Insert a new user into mongo in the convenient way
   *
   * @param username
   *          the username to be set
   * @param password
   *          the passsword in clear text, will be adapted following the definitions of the defined {@link HashStrategy}
   * @param roles
   *          a list of roles to be set
   * @param permissions
   *          a list of permissions to be set
   * @param resultHandler
   *          the ResultHandler will be provided with the id of the generated record
   */
  void insertUser(String username, String password, List<String> roles, List<String> permissions,
      Handler<AsyncResult<String>> resultHandler);

}

File: vertx-auth-mongo/src/main/resources/vertx-auth-mongo-js/mongo_auth.js
/*
 * Copyright 2014 Red Hat, Inc.
 *
 * Red Hat licenses this file to you under the Apache License, version 2.0
 * (the "License"); you may not use this file except in compliance with the
 * License.  You may obtain a copy of the License at:
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */

/** @module vertx-auth-mongo-js/mongo_auth */
var utils = require('vertx-js/util/utils');
var MongoClient = require('vertx-mongo-js/mongo_client');
var HashStrategy = require('vertx-auth-mongo-js/hash_strategy');
var AuthProvider = require('vertx-auth-common-js/auth_provider');

var io = Packages.io;
var JsonObject = io.vertx.core.json.JsonObject;
var JMongoAuth = io.vertx.ext.auth.mongo.MongoAuth;

/**

 @class
*/
var MongoAuth = function(j_val) {

  var j_mongoAuth = j_val;
  var that = this;
  AuthProvider.call(this, j_val);

  /**
   Set the name of the collection to be used. Defaults to DEFAULT_COLLECTION_NAME

   @public
   @param collectionName {string} the name of the collection to be used for storing and reading user data 
   @return {MongoAuth} the current instance itself for fluent calls
   */
  this.setCollectionName = function(collectionName) {
    var __args = arguments;
    if (__args.length === 1 && typeof __args[0] === 'string') {
      j_mongoAuth["setCollectionName(java.lang.String)"](collectionName);
      return that;
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**
   Set the name of the field to be used for the username. Defaults to DEFAULT_USERNAME_FIELD

   @public
   @param fieldName {string} the name of the field to be used 
   @return {MongoAuth} the current instance itself for fluent calls
   */
  this.setUsernameField = function(fieldName) {
    var __args = arguments;
    if (__args.length === 1 && typeof __args[0] === 'string') {
      j_mongoAuth["setUsernameField(java.lang.String)"](fieldName);
      return that;
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**
   Set the name of the field to be used for the password Defaults to DEFAULT_PASSWORD_FIELD

   @public
   @param fieldName {string} the name of the field to be used 
   @return {MongoAuth} the current instance itself for fluent calls
   */
  this.setPasswordField = function(fieldName) {
    var __args = arguments;
    if (__args.length === 1 && typeof __args[0] === 'string') {
      j_mongoAuth["setPasswordField(java.lang.String)"](fieldName);
      return that;
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**
   Set the name of the field to be used for the roles. Defaults to DEFAULT_ROLE_FIELD. Roles are expected to
   be saved as JsonArray

   @public
   @param fieldName {string} the name of the field to be used 
   @return {MongoAuth} the current instance itself for fluent calls
   */
  this.setRoleField = function(fieldName) {
    var __args = arguments;
    if (__args.length === 1 && typeof __args[0] === 'string') {
      j_mongoAuth["setRoleField(java.lang.String)"](fieldName);
      return that;
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**
   Set the name of the field to be used for the permissions. Defaults to DEFAULT_PERMISSION_FIELD.
   Permissions are expected to be saved as JsonArray

   @public
   @param fieldName {string} the name of the field to be used 
   @return {MongoAuth} the current instance itself for fluent calls
   */
  this.setPermissionField = function(fieldName) {
    var __args = arguments;
    if (__args.length === 1 && typeof __args[0] === 'string') {
      j_mongoAuth["setPermissionField(java.lang.String)"](fieldName);
      return that;
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**
   Set the name of the field to be used as property for the username in the method
   {@link AuthProvider#authenticate}. Defaults to DEFAULT_CREDENTIAL_USERNAME_FIELD

   @public
   @param fieldName {string} the name of the field to be used 
   @return {MongoAuth} the current instance itself for fluent calls
   */
  this.setUsernameCredentialField = function(fieldName) {
    var __args = arguments;
    if (__args.length === 1 && typeof __args[0] === 'string') {
      j_mongoAuth["setUsernameCredentialField(java.lang.String)"](fieldName);
      return that;
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**
   Set the name of the field to be used as property for the password of credentials in the method
   {@link AuthProvider#authenticate}. Defaults to DEFAULT_CREDENTIAL_PASSWORD_FIELD

   @public
   @param fieldName {string} the name of the field to be used 
   @return {MongoAuth} the current instance itself for fluent calls
   */
  this.setPasswordCredentialField = function(fieldName) {
    var __args = arguments;
    if (__args.length === 1 && typeof __args[0] === 'string') {
      j_mongoAuth["setPasswordCredentialField(java.lang.String)"](fieldName);
      return that;
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**
   Set the name of the field to be used for the salt. Only used when {@link HashStrategy#setSaltStyle} is
   set to 

   @public
   @param fieldName {string} the name of the field to be used 
   @return {MongoAuth} the current instance itself for fluent calls
   */
  this.setSaltField = function(fieldName) {
    var __args = arguments;
    if (__args.length === 1 && typeof __args[0] === 'string') {
      j_mongoAuth["setSaltField(java.lang.String)"](fieldName);
      return that;
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**
   The name of the collection used to store User objects inside. Defaults to DEFAULT_COLLECTION_NAME

   @public

   @return {string} the collectionName
   */
  this.getCollectionName = function() {
    var __args = arguments;
    if (__args.length === 0) {
      return j_mongoAuth["getCollectionName()"]();
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**
   Get the name of the field to be used for the username. Defaults to DEFAULT_USERNAME_FIELD

   @public

   @return {string} the usernameField
   */
  this.getUsernameField = function() {
    var __args = arguments;
    if (__args.length === 0) {
      return j_mongoAuth["getUsernameField()"]();
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**
   Get the name of the field to be used for the password Defaults to DEFAULT_PASSWORD_FIELD

   @public

   @return {string} the passwordField
   */
  this.getPasswordField = function() {
    var __args = arguments;
    if (__args.length === 0) {
      return j_mongoAuth["getPasswordField()"]();
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**
   Get the name of the field to be used for the roles. Defaults to DEFAULT_ROLE_FIELD. Roles are expected to
   be saved as JsonArray

   @public

   @return {string} the roleField
   */
  this.getRoleField = function() {
    var __args = arguments;
    if (__args.length === 0) {
      return j_mongoAuth["getRoleField()"]();
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**
   Get the name of the field to be used for the permissions. Defaults to DEFAULT_PERMISSION_FIELD.
   Permissions are expected to be saved as JsonArray

   @public

   @return {string} the permissionField
   */
  this.getPermissionField = function() {
    var __args = arguments;
    if (__args.length === 0) {
      return j_mongoAuth["getPermissionField()"]();
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**
   Get the name of the field to be used as property for the username in the method
   {@link AuthProvider#authenticate}. Defaults to DEFAULT_CREDENTIAL_USERNAME_FIELD

   @public

   @return {string} the usernameCredentialField
   */
  this.getUsernameCredentialField = function() {
    var __args = arguments;
    if (__args.length === 0) {
      return j_mongoAuth["getUsernameCredentialField()"]();
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**
   Get the name of the field to be used as property for the password of credentials in the method
   {@link AuthProvider#authenticate}. Defaults to DEFAULT_CREDENTIAL_PASSWORD_FIELD

   @public

   @return {string} the passwordCredentialField
   */
  this.getPasswordCredentialField = function() {
    var __args = arguments;
    if (__args.length === 0) {
      return j_mongoAuth["getPasswordCredentialField()"]();
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**
   Get the name of the field to be used for the salt. Only used when {@link HashStrategy#setSaltStyle} is
   set to 

   @public

   @return {string} the saltField
   */
  this.getSaltField = function() {
    var __args = arguments;
    if (__args.length === 0) {
      return j_mongoAuth["getSaltField()"]();
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**
   The HashStrategy which is used by the current instance

   @public
   @param hashStrategy {HashStrategy} the {@link HashStrategy} to be set 
   @return {MongoAuth} the current instance itself for fluent calls
   */
  this.setHashStrategy = function(hashStrategy) {
    var __args = arguments;
    if (__args.length === 1 && typeof __args[0] === 'object' && __args[0]._jdel) {
      j_mongoAuth["setHashStrategy(io.vertx.ext.auth.mongo.HashStrategy)"](hashStrategy._jdel);
      return that;
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**
   The HashStrategy which is used by the current instance

   @public

   @return {HashStrategy} the defined instance of {@link HashStrategy}
   */
  this.getHashStrategy = function() {
    var __args = arguments;
    if (__args.length === 0) {
      return utils.convReturnVertxGen(j_mongoAuth["getHashStrategy()"](), HashStrategy);
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**
   Insert a new user into mongo in the convenient way

   @public
   @param username {string} the username to be set 
   @param password {string} the passsword in clear text, will be adapted following the definitions of the defined {@link HashStrategy} 
   @param roles {Array.<string>} a list of roles to be set 
   @param permissions {Array.<string>} a list of permissions to be set 
   @param resultHandler {function} the ResultHandler will be provided with the id of the generated record 
   */
  this.insertUser = function(username, password, roles, permissions, resultHandler) {
    var __args = arguments;
    if (__args.length === 5 && typeof __args[0] === 'string' && typeof __args[1] === 'string' && typeof __args[2] === 'object' && __args[2] instanceof Array && typeof __args[3] === 'object' && __args[3] instanceof Array && typeof __args[4] === 'function') {
      j_mongoAuth["insertUser(java.lang.String,java.lang.String,java.util.List,java.util.List,io.vertx.core.Handler)"](username, password, roles, permissions, function(ar) {
      if (ar.succeeded()) {
        resultHandler(ar.result(), null);
      } else {
        resultHandler(null, ar.cause());
      }
    });
    } else throw new TypeError('function invoked with invalid arguments');
  };

  // A reference to the underlying Java delegate
  // NOTE! This is an internal API and must not be used in user code.
  // If you rely on this property your code is likely to break if we change it / remove it without warning.
  this._jdel = j_mongoAuth;
};

/**
 Creates an instance of MongoAuth by using the given  and configuration object. An example for a
 configuration object:
 
 <pre>
 JsonObject js = new JsonObject();
 js.put(MongoAuth.PROPERTY_COLLECTION_NAME, createCollectionName(MongoAuth.DEFAULT_COLLECTION_NAME));
 </pre>

 @memberof module:vertx-auth-mongo-js/mongo_auth
 @param mongoClient {MongoClient} an instance of  to be used for data storage and retrival 
 @param config {Object} the configuration object for the current instance. By this 
 @return {MongoAuth} the created instance of {@link MongoAuth}s
 */
MongoAuth.create = function(mongoClient, config) {
  var __args = arguments;
  if (__args.length === 2 && typeof __args[0] === 'object' && __args[0]._jdel && (typeof __args[1] === 'object' && __args[1] != null)) {
    return utils.convReturnVertxGen(JMongoAuth["create(io.vertx.ext.mongo.MongoClient,io.vertx.core.json.JsonObject)"](mongoClient._jdel, utils.convParamJsonObject(config)), MongoAuth);
  } else throw new TypeError('function invoked with invalid arguments');
};

// We export the Constructor function
module.exports = MongoAuth;